### ISSUE TITLE
Copying data to target device throw "maximum recursion depth exceeded"

### ISSUE THREAD

**Issue by fatevase** [open-mmlab/mmengine#575]
Thanks for your error report and we appreciate it a lot.
**Describe the bug**
I try save some str type data in data_sampler on dataloader. but str type data will cause a recursion error.

**Reproduction**

1. What command or script did you run?
custom dataset.

```none
data_list.append(# include str
    dict(img=anno[1].reshape(32,32,3), img_label=anno[0], name=anno[2].decode('utf-8')) 
)
return dict(metainfo=metainfo, data_list=data_list)
# pipeline will map name key and value to data_sampler dict.
```

**Bug fix**
origin file: model.base_model.data_preprocessor.py,  sequence include str, list, tuple, str will keep loop in cast_data.

```python
    def cast_data(self, data: CastData) -> CastData:
        if isinstance(data, Mapping):
            return {key: self.cast_data(data[key]) for key in data}
        elif isinstance(data, tuple) and hasattr(data, '_fields'):
            # namedtuple
            return type(data)(*(self.cast_data(sample)for sample in data))  # type: ignore  # noqa: E501  # yapf:disable
        elif isinstance(data, Sequence): 
            return [self.cast_data(sample) for sample in data]
        elif isinstance(data, torch.Tensor):
            return data.to(self.device)
        elif isinstance(data, BaseDataElement):
            return data.to(self.device)
        else:            
            return data
```

I think the cast_data() may want to loop through all structures and change the data that can be put into the gpu, but the loop is still triggered when the data is str.
I change `elif isinstance(data, Sequence): ` to `elif isinstance(data, (list, tuple)): `.

**Comment by HAOCHENYE** [comment]
Thanks for your feed back! We'll fix it ASAP!

**Comment by HAOCHENYE** [comment]
Hi! I solved this issue in #602 :wink:
